Mathlib/LinearAlgebra/TensorProduct/Submodule.lean

```diff
@@ -0,0 +1,253 @@
+`i(R) ⊗[R] N` and `N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This generalizes `TensorProduct.lid` as `i(R)` is not necessarily isomorphic to `R`. -/
+def lTensorOne : (1 : Submodule R S) ⊗[R] N ≃ₗ[R] N := by
+  refine LinearEquiv.ofLinear N.lTensorOne' (TensorProduct.mk R (1 : Submodule R S) N
+    ⟨1, one_le.1 (le_refl _)⟩) (by ext; simp [lTensorOne']) (TensorProduct.ext' fun r n ↦ ?_)
+  change ⟨1, _⟩ ⊗ₜ[R] lTensorOne' N _ = r ⊗ₜ[R] n
+  obtain ⟨x, y, h : algebraMap R S y = x⟩ := r
+  simp_rw [← h, lTensorOne'_apply, ← TensorProduct.smul_tmul,
+    SetLike.mk_smul_mk, Algebra.smul_def, mul_one]
+
+theorem lTensorOne_apply
+    (y : R) {hy : algebraMap R S y ∈ (1 : Submodule R S)}
+    (n : N) : N.lTensorOne (⟨algebraMap R S y, hy⟩ ⊗ₜ[R] n) = y • n :=
+  N.lTensorOne'_apply y n
+
+theorem lTensorOne_symm_apply (n : N) :
+    N.lTensorOne.symm n = ⟨1, one_le.1 (le_refl _)⟩ ⊗ₜ[R] n := rfl
+
+/-- If `M` is a submodule in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] i(R) →ₗ[R] M` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This is promoted to an isomorphism as `Submodule.rTensorOne`. Use that instead. -/
+def rTensorOne' : M ⊗[R] (1 : Submodule R S) →ₗ[R] M :=
+  (LinearEquiv.ofEq _ _ (by rw [mulMap_range, mul_one])).toLinearMap ∘ₗ (mulMap M _).rangeRestrict
+
+theorem rTensorOne'_apply
+    (y : R) {hy : algebraMap R S y ∈ (1 : Submodule R S)}
+    (m : M) : M.rTensorOne' (m ⊗ₜ[R] ⟨algebraMap R S y, hy⟩) = y • m :=
+  Subtype.val_injective <| by simp [rTensorOne', Algebra.smul_def, Algebra.commutes y m.1]
+
+/-- If `M` is a submodule in an algebra `S` over `R`, there is the natural isomorphism between
+`M ⊗[R] i(R)` and `M` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This generalizes `TensorProduct.rid` as `i(R)` is not necessarily isomorphic to `R`. -/
+def rTensorOne : M ⊗[R] (1 : Submodule R S) ≃ₗ[R] M := by
+  refine LinearEquiv.ofLinear M.rTensorOne' ((TensorProduct.comm R _ _).toLinearMap ∘ₗ
+    TensorProduct.mk R (1 : Submodule R S) M ⟨1, one_le.1 (le_refl _)⟩)
+      (by ext; simp [rTensorOne']) (TensorProduct.ext' fun n r ↦ ?_)
+  change rTensorOne' M _ ⊗ₜ[R] ⟨1, _⟩ = n ⊗ₜ[R] r
+  obtain ⟨x, y, h : algebraMap R S y = x⟩ := r
+  simp_rw [← h, rTensorOne'_apply, TensorProduct.smul_tmul,
+    SetLike.mk_smul_mk, Algebra.smul_def, mul_one]
+
+theorem rTensorOne_apply
+    (y : R) {hy : algebraMap R S y ∈ (1 : Submodule R S)}
+    (m : M) : M.rTensorOne (m ⊗ₜ[R] ⟨algebraMap R S y, hy⟩) = y • m :=
+  M.rTensorOne'_apply y m
+
+theorem rTensorOne_symm_apply (m : M) :
+    M.rTensorOne.symm m = m ⊗ₜ[R] ⟨1, one_le.1 (le_refl _)⟩ := rfl
+
+variable {M} in
+/-- If `m : ι → M` is a family of elements, then there is an `R`-linear map from `ι →₀ N` to
+`S` which maps `{ n_i }` to the sum of `m_i * n_i`. -/
+def mulLeftMap {ι : Type*} (m : ι → M) : (ι →₀ N) →ₗ[R] S :=
```

Comment:
I think `Finsupp.lsum R fun i ↦ (m i).1 • N.subtype` is a more intuitive definition. Is there a reason why this definition is used?

Reply:
And similarly, `Finsupp.lsum R fun i ↦ MulOpposite.op (n i).1 • M.subtype` for `mulRightMap`.

Reply:
Let me try.

Reply:
Ah, I remembered the reason: I need to express `mulLeftMap` as a composition of `mulMap` and a linear map which is injective or an isomorphism in the proof of equivalency of different definitions of linearly disjoint. So this property is essential.

But maybe we can use the new definition and state the old definition as `mulLeftMap_eq_mulMap_comp`?

Reply:
And maybe the definition of these can go into the same file as `Finsupp.lsum`?

Reply:
> maybe the definition of these can go into the same file as `Finsupp.lsum`?

That cannot be done since `Algebra` is not imported in `Mathlib.LinearAlgebra.Finsupp`.

Reply:
Maybe
```
import Mathlib.LinearAlgebra.Finsupp

variable {R R' S} [CommSemiring R] [CommSemiring R'] [AddCommMonoid S]
  [Module R S] [Module R' S]
  [SMulCommClass R R' S] {N : Submodule R S}

noncomputable
def smulLeftSubtype {ι : Type*} (m : ι → R') : (ι →₀ N) →ₗ[R] S :=
  Finsupp.lsum R fun i ↦ (m i) • N.subtype
```

Reply:
Thanks. I found that `[Algebra R S]` could be replace by `[Module R S] [SMulCommClass R S S]` So probably I can define `Submodule.mulLeftMap` in `Mathlib.LinearAlgebra.Finsupp`. Do you think your generalization `smulLeftSubtype` is still useful?